contributors/design-proposals/config-conventions.md

```diff
@@ -0,0 +1,222 @@
+1.  API features: Use (or implement) functionality of the API (e.g. think through and implement initializers instead of --register-with-label). Allowing for extension in the right places is a better way to give users control.
+1.  Feature discovery: Write components that introspect the existing API to decide whether to enable a feature or not. E.g. controller-manager should start an app controller if the app API is available, kubelet should enable zram if zram is set in the node spec.
+1.  Downwards API: Use the APIs that the OS and pod environment expose directly before opting to pass in new configuration options.
+1.  const's: If you don't know whether tweaking a value will be useful, make the value const. Only give it a configuration option once there becomes a need to tweak the value at runtime.
+1.  Autotuning: Build systems that incorporate feedback and do the best thing under the given circumstances. This makes the system more robust. (e.g. prefer congestion control, load shedding, backoff rather than explicit limiting).
+1.  Avoid feature flags: Turn on features when they are tested and ready for production. Don't use feature flags as a fallback for poorly tested code.
+1.  Configuration profiles: Instead of allowing individual configuration options to be modified, try to encompass a broader desire as a configuration profile. For example: instead of enabling individual alpha features, have an EnableAlpha option that enables all. Instead of allowing individual controller knobs to be modified, have a TestMode option that sets a broad number of parameters to be suitable for tests.
+
+## Part 2: Component Configuration
+
+### Versioning Configuration
+
+We create configuration API groups per component that live in the source tree
+of the component. Each component has its own API group for configuration.
+Components will use the same API machinery that we use for other API groups.
+Configuration API serialization doesn't have the same performance requirements
+as other APIs so much of the codegen can be avoided (e.g. ugorji, generated
+conversions) and we can instead fallback to the reflection based
+implementations where they exist.
+
+
+### Retrieving Configuration
+
+The primary mechanism for retrieving static configuration should be
+deserialization from files. For the majority of components (with the possible
+exception of the kubelet, see
+[here](https://github.com/kubernetes/kubernetes/pull/29459)), these files will
+be source from the configmap API and managed by the kubelet. Reliability of
+this mechanism is predicated on kubelet checkpointing of pod dependencies.
+
+
+### Structuring Configuration
+
+Group related options into distinct objects and subobjects. Instead of writing:
+
+
+```yaml
+kind: KubeProxyConfiguration
+apiVersion: kubeproxy/v1beta3
+ipTablesSyncPeriod: 2
+ipTablesConntrackHashSize: 2
+ipTablesConntrackTableSize: 2
+```
+
+Write:
+
+```yaml
+kind: KubeProxyConfiguration
+apiVersion: kubeproxy/v1beta3
```

Comment:
There is discussion on the SIG architecture thread about changing the API group naming convention.

We have componentconfig, kubeletconfig, and probably other variants currently.

My proposal was `<component>.config.k8s.io` (assuming we're going with .k8s.io suffixes).